### PR TITLE
Graduate JobPodFailurePolicy to stable

### PR DESCRIPTION
**Pull request by mimowo**
#### What type of PR is this?

/kind feature
/kind api-change
/sig apps

#### Which issue(s) this PR fixes:

Tracking issue: https://github.com/kubernetes/enhancements/issues/3329

#### Special notes for your reviewer:

The related code PRs for the graduation:
- https://github.com/kubernetes/kubernetes/pull/125461
- https://github.com/kubernetes/kubernetes/pull/125442
- https://github.com/kubernetes/kubernetes/pull/125482

The PRs are split for ease of reviewing as the changes are independent and may require different expertise / reviewers.

#### Does this PR introduce a user-facing change?

```release-note
Graduate JobPodFailurePolicy to GA and lock
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures
```

### CODE REVIEW

**Comment by mimowo** [comment]
/remove-sig node
/remove-area kubelet
After the split of PodDisruptionConditions to https://github.com/kubernetes/kubernetes/pull/125461

**Comment by mimowo** [comment]
/test pull-kubernetes-node-e2e-containerd 
it is ticketed already: https://github.com/kubernetes/kubernetes/issues/123852

**Comment by mimowo** [comment]
/test pull-kubernetes-e2e-gce-cos-alpha-features
restest, seems infra issues, there is no build log even

File: pkg/controller/job/job_controller_test.go [hunk 1]
```diff
@@ -2593,19 +2588,18 @@ func TestSyncJobWithJobPodFailurePolicy(t *testing.T) {
...
-		enableJobPodReplacementPolicy bool
-		job                           batch.Job
-		pods                          []v1.Pod
-		wantConditions                *[]batch.JobCondition
-		wantStatusFailed              int32
-		wantStatusActive              int32
-		wantStatusSucceeded           int32
-		wantStatusTerminating         *int32
+		enablePodDisruptionConditions      bool
+		TestSyncJobWithJobPodFailurePolicy bool
```
Comment (dejanzele) [inline pkg/controller/job/job_controller_test.go:2592]: This field doesn't seem to be used anywhere

Comment (mimowo) [reply pkg/controller/job/job_controller_test.go]: Right, not sure it got there, but removed now :)

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Comment by atiratree** [comment]
lgtm, also I presume the code gating changes will be done after this PR and https://github.com/kubernetes/kubernetes/pull/125461 are merged, right?

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve

**Comment by mimowo** [comment]
> lgtm, also I presume the code gating changes will be done after this PR and #125461 are merged, right?

Actually, I was thinking we merge the PR as is (combining the obsolete test removal, and feature gate update) as the feature gate is a "test" that we didn't forget to update a test. This approach was taken in the past for example: https://github.com/kubernetes/kubernetes/pull/123575/, or https://github.com/kubernetes/kubernetes/pull/121302/.

**Comment by mimowo** [comment]
Rebased after https://github.com/kubernetes/kubernetes/pull/125461 got merged.

**Comment by mimowo** [comment]
`pull-kubernetes-typecheck` and `pull-kubernetes-verify-lint` complained about the extra field `jobPodFailurePolicy` which was used in a new test cases added in https://github.com/kubernetes/kubernetes/pull/125985. I rebased and removed the field use. Please renew lgtm.

**Comment by mimowo** [comment]
/test pull-kubernetes-integration 
overloaded infra

**Comment by mimowo** [comment]
One more rebase after https://github.com/kubernetes/kubernetes/pull/126030

**Comment by mimowo** [comment]
One more rebase after https://github.com/kubernetes/kubernetes/pull/125510. The conflicts were only inside job_controller_test.go, due to new feature gates fields added in this PR.

**Pull request merged**